### PR TITLE
Sine basis

### PR DESCRIPTION
**Pull request by AramisDuf**
Fixing bugs in trend and in "multiplying" custom basis. Inmplementing the standing wave basis.

### CODE REVIEW

File: src/spcqe/functions.py [hunk 1]
```diff
@@ -3,46 +3,64 @@
 from itertools import combinations
 
 
-def make_basis_matrix(num_harmonics, length, periods, trend=False, max_cross_k=None, custom_basis=None):
+def make_basis_matrix(num_harmonics, length, periods, standing_wave=False, trend=False, max_cross_k=None, custom_basis=None):
+
```
Comment (AramisDuf) [inline src/spcqe/functions.py:7]: Make sphinx compliant docstring

File: src/spcqe/functions.py [hunk 2]
```diff
@@ -52,27 +70,46 @@ def make_basis_matrix(num_harmonics, length, periods, trend=False, max_cross_k=N
...
     B = np.hstack(B_list)
     return B
 
-
-def make_regularization_matrix(num_harmonics, weight, periods, trend=False, max_cross_k=None, custom_basis=None):
-    num_harmonics = np.atleast_1d(num_harmonics)
+# TODO: is it different if standing wave is True?
+def make_regularization_matrix(num_harmonics, weight, periods, standing_wave=False, trend=False, max_cross_k=None, custom_basis=None):
+    
+    # Sanity checks
```
Comment (AramisDuf) [inline src/spcqe/functions.py:83]: make it a helper function

File: src/spcqe/functions.py [hunk 3]
```diff
@@ -52,27 +70,46 @@ def make_basis_matrix(num_harmonics, length, periods, trend=False, max_cross_k=N
...
     B_list = B0 + B_fourier + C
     B = np.hstack(B_list)
     return B
 
-
-def make_regularization_matrix(num_harmonics, weight, periods, trend=False, max_cross_k=None, custom_basis=None):
-    num_harmonics = np.atleast_1d(num_harmonics)
+# TODO: is it different if standing wave is True?
+def make_regularization_matrix(num_harmonics, weight, periods, standing_wave=False, trend=False, max_cross_k=None, custom_basis=None):
+
```
Comment (AramisDuf) [inline src/spcqe/functions.py:82]: docstring

File: src/spcqe/functions.py [hunk 4]
```diff
@@ -52,27 +70,46 @@ def make_basis_matrix(num_harmonics, length, periods, trend=False, max_cross_k=N
...
 
     B_list = B0 + B_fourier + C
     B = np.hstack(B_list)
     return B
 
-
-def make_regularization_matrix(num_harmonics, weight, periods, trend=False, max_cross_k=None, custom_basis=None):
-    num_harmonics = np.atleast_1d(num_harmonics)
+# TODO: is it different if standing wave is True?
+def make_regularization_matrix(num_harmonics, weight, periods, standing_wave=False, trend=False, max_cross_k=None, custom_basis=None):
```
Comment (AramisDuf) [inline src/spcqe/functions.py:81]: test standing wave vs non standing wave

**Review by AramisDuf — COMMENTED** [review]
Completed the code review with Bennet

**Review by bmeyers — CHANGES_REQUESTED** [review]
Approved after comments above addressed

**Comment by AramisDuf** [comment]
Resolved all the comments

File: src/spcqe/functions.py [hunk 5]
```diff
@@ -1,39 +1,59 @@
+"""
+This module contains functions to create basis and
+regularization matrices for a time series with one or
+multiple periods.
+"""
```
Comment (bmeyers) [inline src/spcqe/functions.py:5]: add your name as author

File: src/spcqe/functions.py [hunk 6]
```diff
@@ -1,39 +1,59 @@
...
+    max_cross_k=None,
+    custom_basis=None,
+):
+    """
+    This function creates a basis matrix for a time series with one or multiple
+    periods. It creates a Fourier basis matrix for each period, then combines
+    them into a single matrix, including cross-terms. It also includes an offset
+    term, and can include a linear trend.
+
+    :param num_harmonics: int or ndarray, 1D array of ints, number of harmonics to
```
Comment (bmeyers) [inline src/spcqe/functions.py:27]: move type to separate entries `:type arg:` style

File: src/spcqe/functions.py [hunk 7]
```diff
@@ -69,48 +87,59 @@ def make_basis_matrix(num_harmonics, length, periods, standing_wave=False, trend
...
+    max_cross_k=None,
+    custom_basis=None,
+):
+    """
+    This function creates a regularization matrix for a time series with one or
+    multiple periods. It creates a regularization matrix for each period, then
+    combines them into a single matrix. The weights are determined using the
+    Dirichlet energy of the basis functions.
+
+    :param num_harmonics: int or ndarray, 1D array of ints, number of harmonics to
```
Comment (bmeyers) [inline src/spcqe/functions.py:116]: same as above, add types in separate lines

**Review by bmeyers — APPROVED** [review]

**Pull request merged**